pgbouncer-krb/entrypoint.sh

```diff
@@ -0,0 +1,205 @@
+  whoami
+  ls -la /etc/krb5*
+  tee /etc/krb5.conf <<EOF
+[libdefaults]
+  rdns = false
+  default_realm = $REALM
+[realms]
+  $REALM = {
+    kdc = kdc-kadmin
+    admin_server = kdc-kadmin
+  }
+EOF
+  echo ""
+
+  echo "==================================================================================="
+  echo "==== Testing ======================================================================"
+  echo "==================================================================================="
+  until kadminCommand "list_principals $KADMIN_PRINCIPAL_FULL"; do
+    >&2 echo "KDC is unavailable - sleeping 1 sec"
+    sleep 1
+  done
+  echo "KDC and Kadmin are operational"
+  echo ""
+fi
+
+# Here are some parameters. See all on
+# https://pgbouncer.github.io/config.html
+
+PG_CONFIG_DIR=/etc/pgbouncer
+
+if [ -n "$DATABASE_URL" ]; then
+  # Thanks to https://stackoverflow.com/a/17287984/146289
+
+  # Allow to pass values like dj-database-url / django-environ accept
+  proto="$(echo "$DATABASE_URL" | grep :// | sed -e's,^\(.*://\).*,\1,g')"
+  url="$(echo "$DATABASE_URL" | sed -e s,"$proto",,g)"
+
+  # extract the user and password (if any)
+  userpass=$(echo "$url" | grep @ | sed -r 's/^(.*)@([^@]*)$/\1/')
+  DB_PASSWORD="$(echo "$userpass" | grep : | cut -d: -f2)"
+  if [ -n "$DB_PASSWORD" ]; then
+    DB_USER=$(echo "$userpass" | grep : | cut -d: -f1)
+  else
+    DB_USER=$userpass
+  fi
+
+  # extract the host -- updated
+  hostport=$(echo "$url" | sed -e s,"$userpass"@,,g | cut -d/ -f1)
+  port=$(echo "$hostport" | grep : | cut -d: -f2)
+  if [ -n "$port" ]; then
+    DB_HOST=$(echo "$hostport" | grep : | cut -d: -f1)
+      DB_PORT=$port
+  else
+      DB_HOST=$hostport
+  fi
+
+  DB_NAME="$(echo "$url" | grep / | cut -d/ -f2-)"
+fi
+
+# Write the password with MD5 encryption, to avoid printing it during startup.
+# Notice that `docker inspect` will show unencrypted env variables.
+_AUTH_FILE="${AUTH_FILE:-$PG_CONFIG_DIR/userlist.txt}"
+
+# Workaround userlist.txt missing issue
+# https://github.com/edoburu/docker-pgbouncer/issues/33
+if [ ! -e "${_AUTH_FILE}" ]; then
+  touch "${_AUTH_FILE}"
+fi
+# shellcheck disable=SC2166
+if [ -n "$DB_USER" -a -n "$DB_PASSWORD" -a -e "${_AUTH_FILE}" ] && ! grep -q "^\"$DB_USER\"" "${_AUTH_FILE}"; then
+  if [ "$AUTH_TYPE" != "plain" ]; then
+     pass="md5$(echo "$DB_PASSWORD$DB_USER" | md5sum | cut -f 1 -d ' ')"
+  else
+     pass="$DB_PASSWORD"
+  fi
+  echo "\"$DB_USER\" \"$pass\"" >> ${PG_CONFIG_DIR}/userlist.txt
+  echo "Wrote authentication credentials to ${PG_CONFIG_DIR}/userlist.txt"
+fi
+
+if [ ! -f ${PG_CONFIG_DIR}/pgbouncer.ini ]; then
+  echo "Create pgbouncer config in ${PG_CONFIG_DIR}"
+
+# Config file is in "ini" format. Section names are between "[" and "]".
+# Lines starting with ";" or "#" are taken as comments and ignored.
+# The characters ";" and "#" are not recognized when they appear later in the line.
+# shellcheck disable=SC2059
+  printf "\
```

Comment:
This would be easier to maintain as a HEREDOC:
```suggestion
cat <"EOF" > "${PG_CONFIG_DIR}/pgbouncer.ini"
```
then strip out the line endings and just do it exactly like you'd see in the config with shell variables interpolated.

Reply:
fixed in 686abd4

Reply:
Oh, I think I see now why they did it with `printf`. Using `printf` (or `echo -en`), you have to specify the line ending, which lets you skip the line ending if the var is not printed. That might be worth undoing. The way it is now, the config file will be full of empty lines for any vars that are not defined.

Reply:
Yeah, heredoc has the fewest empty lines:

```
#!/usr/bin/env bash

echo "=========== cat syntax"
cat <<EOF
${FOO:+foo = ${FOO}}
${SHELL:+shell = ${SHELL}}
${BAZ:+baz = ${BAZ}}
${HOME:+home = ${HOME}}
EOF

echo "=========== echo syntax"
echo -en "
${FOO:+foo = ${FOO}\n}
${SHELL:+shell = ${SHELL}\n}
${BAZ:+baz = ${BAZ}\n}
${HOME:+home = ${HOME}\n}"

echo "=========== printf syntax"
printf "
${FOO:+foo = ${FOO}\n}
${SHELL:+shell = ${SHELL}\n}
${BAZ:+baz = ${BAZ}\n}
${HOME:+home = ${HOME}\n}"
$ ./test.sh
=========== cat syntax

shell = /opt/homebrew/bin/bash

home = /Users/danielh
=========== echo syntax


shell = /opt/homebrew/bin/bash


home = /Users/danielh
=========== printf syntax


shell = /opt/homebrew/bin/bash


home = /Users/danielh
```

Reply:
yeah it's okey. Big win when using `cat <<EOF` i removed ignore shellcheck `disable=SC2059`.
 So now we enable all checks. Which is good.